x/pairing/keeper/msg_server_freeze.go

```diff
@@ -0,0 +1,35 @@
+package keeper
+
+import (
+	"context"
+	"math"
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/utils"
+	epochstoragetypes "github.com/lavanet/lava/x/epochstorage/types"
+	"github.com/lavanet/lava/x/pairing/types"
+)
+
+func (k msgServer) Freeze(goCtx context.Context, msg *types.MsgFreeze) (*types.MsgFreezeResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	providerAddr, err := sdk.AccAddressFromBech32(msg.GetCreator())
+	if err != nil {
+		return nil, utils.LavaFormatError("Freeze_get_provider_address", err, &map[string]string{"providerAddress": msg.GetCreator()})
+	}
+
+	for _, chainId := range msg.GetChainIds() {
+		stakeEntry, found, index := k.epochStorageKeeper.GetStakeEntryByAddressCurrent(ctx, epochstoragetypes.ProviderKey, chainId, providerAddr)
+		if !found {
+			continue
+		}
+
+		// freeze the provider by making the StakeAppliedBlock be max. This will remove the provider from the pairing list in the next epoch
+		stakeEntry.StakeAppliedBlock = math.MaxUint64
```

Comment:
This may interfere payments to this provider (for services done before the freeze) from going through while frozen. At least if we allow unstaking of a frozen provider, and may also for "regular" staked frozen providers.